src/inmanta/util/async_lru.py

```diff
@@ -0,0 +1,73 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+
+    based on code from https://github.com/python/cpython/issues/90780
+"""
+
+import asyncio
+from asyncio import Future, Lock
+
+# Async support for @functools.lrucache
+# From https://github.com/python/cpython/issues/90780
+from functools import _lru_cache_wrapper, lru_cache, wraps
+from typing import Any, Awaitable, Callable, Generator, Literal, ParamSpec, TypeVar, overload
+
+T = TypeVar("T")
+_PWrapped = ParamSpec("_PWrapped")
+
+
+class CachedAwaitable(Awaitable[T]):
+    def __init__(self, awaitable: Awaitable[T]) -> None:
+        self.awaitable = awaitable
+        self.result: Future[T] | None = None
+
+    def __await__(self) -> Generator[Any, None, T]:
+        if self.result is None:
+            fut = asyncio.get_event_loop().create_future()
+            self.result = fut
+            self.result = yield from self.awaitable.__await__()
+            fut.set_result(self.result)
```

Comment:
This part is not like it was on https://github.com/python/cpython/issues/90780

it didn't work as proposed when the cache was hit before the awaitable was done.

I only half understand this code, but it seems to work